File: addons/terrabrush/src/Scripts/KeybindSettings.cs
#if TOOLS
using Godot;
using System.Text;

namespace TerraBrush;

[Tool]
public partial class KeybindSettings : AcceptDialog {
	[NodePath] private Tree _keybindList;
	private TreeItem _root;
	private KeybindManager _keybindManager;

	private enum ShortcutType {
	    Add = 1,
	    Erase  = 2
	}

	private string ProperCase(StringName name) {
		var parts = name.ToString().Split("_");
		var newString = new StringBuilder();
		foreach (var part in parts) {
			if (part != "terrabrush") {
			    newString.Append(char.ToUpperInvariant(part[0]) + part[1..]);
			    newString.Append(' ');
			}
		}

		return newString.ToString();
	}

	// Called when the node enters the scene tree for the first time.
	public override void _Ready() {
		this.RegisterNodePaths();
		var iconTheme = EditorInterface.Singleton.GetEditorTheme();
		_keybindManager = new KeybindManager();
		_root = _keybindList.CreateItem();
		_keybindList.SetColumnTitle(0,"Name");
		_keybindList.SetColumnTitle(1, "Binding");
		_keybindList.HideRoot = true;
		foreach (var action in _keybindManager.ActionNames) {
			var iter = _keybindList.CreateItem(_root);
			iter.SetText(0, ProperCase(action));
			iter.SetText(1, _keybindManager.DescribeKey(action));

			iter.AddButton(1, iconTheme.GetIcon("Edit", "EditorIcons"), (int)ShortcutType.Add);
			iter.AddButton(1, iconTheme.GetIcon("Close", "EditorIcons"), (int)ShortcutType.Erase);
			iter.SetMetadata(0, action);
		}
		_keybindList.ButtonClicked += KeybindListOnButtonClicked;
	}

	private void KeybindListOnButtonClicked(TreeItem item, long column, long id, long mouseButtonIndex) {
		if (column != 1) return;
		if ((MouseButton)mouseButtonIndex != MouseButton.Left) return;

		switch ((ShortcutType)id) {
			case ShortcutType.Add:
				var dlg = ResourceLoader.Load<PackedScene>("res://addons/terrabrush/Components/KeyListenDialog.tscn")
					.Instantiate<KeyListenDialog>();

				dlg.KeyListenAccepted += (key) => {
					var action = item.GetMetadata(0).AsStringName();
					item.SetText(1, KeybindManager.DescribeKey(key));
					_keybindManager.UpdateKeybind(action, key);
					dlg.QueueFree();
				};

				dlg.KeyListenCancelled += () => dlg.QueueFree();

				GetTree().Root.AddChild(dlg);
				dlg.PopupCentered();
				// Handle Erase Shortcut
				break;
			case ShortcutType.Erase:
				// Handle Add Shortcut
				var action = item.GetMetadata(0).AsStringName();
				_keybindManager.ResetKeybind(action);
				item.SetText(1, _keybindManager.DescribeKey(action));
				break;
		}
	}
}
#endif

File: src/editor_nodes/keybind_settings.cpp
#include "keybind_settings.h"
#include "key_listen_dialog.h"

#include <godot_cpp/classes/editor_interface.hpp>
#include <godot_cpp/classes/theme.hpp>
#include <godot_cpp/classes/tree.hpp>
#include <godot_cpp/classes/v_box_container.hpp>
#include <godot_cpp/classes/input_event_key.hpp>
#include <godot_cpp/classes/scene_tree.hpp>

using namespace godot;

void KeybindSettings::_bind_methods() {
    ClassDB::bind_method(D_METHOD("keybindListOnButtonClicked", "item", "column", "id", "mouseButtonIndex"), &KeybindSettings::keybindListOnButtonClicked);
    ClassDB::bind_method(D_METHOD("onConfirm"), &KeybindSettings::onConfirm);
}

KeybindSettings::KeybindSettings() {}

KeybindSettings::~KeybindSettings() {}

void KeybindSettings::_ready() {
    set_ok_button_text("Close");
    set_title("TerraBrush Keybindings");
    set_initial_position(WindowInitialPosition::WINDOW_INITIAL_POSITION_CENTER_MAIN_WINDOW_SCREEN);
    set_size(Vector2(400, 300));
    set_visible(true);
    connect("confirmed", Callable(this, "onConfirm"));

    VBoxContainer *vboxContainer = memnew(VBoxContainer);
    vboxContainer->set_anchors_and_offsets_preset(Control::LayoutPreset::PRESET_FULL_RECT);
    add_child(vboxContainer);

    Tree *tree = memnew(Tree);
    tree->set_columns(2);
    tree->set_column_titles_visible(true);
    tree->set_hide_folding(true);
    tree->set_hide_root(true);
    tree->set_v_size_flags(Control::SizeFlags::SIZE_EXPAND_FILL);
    vboxContainer->add_child(tree);

    Ref<Theme> iconTheme = EditorInterface::get_singleton()->get_editor_theme();

    _keybindManager = Ref<KeybindManager>(memnew(KeybindManager));

    TreeItem *root = tree->create_item();
    tree->set_column_title(0,"Name");
    tree->set_column_title(1, "Binding");
    tree->set_hide_root(true);
    for (String action : _keybindManager->get_actionNames()) {
        TreeItem *iter = tree->create_item(root);
        iter->set_text(0, properCase(action));
        iter->set_text(1, _keybindManager->describeKey(action));

        iter->add_button(1, iconTheme->get_icon("Edit", "EditorIcons"), (int)ShortcutType::SHORTCUTTYPE_ADD);
        iter->add_button(1, iconTheme->get_icon("Close", "EditorIcons"), (int)ShortcutType::SHORTCUTTYPE_ERASE);
        iter->set_metadata(0, action);
    }
    tree->connect("button_clicked", Callable(this, "keybindListOnButtonClicked"));
}

String KeybindSettings::properCase(StringName name) {
    PackedStringArray parts = name.split("_");
    String newString = "";
    for (String part : parts) {
        if (part != "terrabrush") {
            newString += part.substr(0, 1).to_upper() + part.substr(1, part.length() - 1);
            newString += " ";
        }
    }

    return newString;
}

void KeybindSettings::keybindListOnButtonClicked(const TreeItem *item, const int column, const int id, const int mouseButtonIndex) {
    if (column != 1) return;
    if ((MouseButton)mouseButtonIndex != MouseButton::MOUSE_BUTTON_LEFT) return;

    switch ((ShortcutType)id) {
        case ShortcutType::SHORTCUTTYPE_ADD: {
            KeyListenDialog *dlg = memnew(KeyListenDialog);
            dlg->init(([dlg, item, &_keybindManager = _keybindManager](Ref<InputEventKey> key) {
                String action = item->get_metadata(0);
                const_cast<TreeItem*>(item)->set_text(1, KeybindManager::describeInputKey(key));
                _keybindManager->updateKeybind(action, key);
                dlg->queue_free();
            }));

            get_tree()->get_root()->add_child(dlg);
            dlg->popup_centered();
            break;
        }
        case ShortcutType::SHORTCUTTYPE_ERASE:
            String action = item->get_metadata(0);
            _keybindManager->resetKeybind(action);
            const_cast<TreeItem*>(item)->set_text(1, _keybindManager->describeKey(action));
            break;
    }
}

void KeybindSettings::onConfirm() {
    _keybindManager->saveEditorSettings();
    queue_free();
}